widget/text.go

```diff
@@ -168,13 +168,21 @@ func (t *Text) draw(screen *ebiten.Image) {
 	case TextPositionCenter:
 		p = p.Add(image.Point{0, int((float64(r.Dy()) - t.measurements.boundingBoxHeight) / 2)})
 	case TextPositionEnd:
-		p = p.Add(image.Point{0, int((float64(r.Dy()) - t.measurements.boundingBoxHeight))})
+		p = p.Add(image.Point{0, int(float64(r.Dy()) - t.measurements.boundingBoxHeight)})
 	}
 
 	t.colorList = &datastructures.Stack[color.Color]{}
 	t.colorList.Push(&t.Color)
 
 	for i, line := range t.measurements.lines {
+		ly := int(math.Round(float64(p.Y) + t.measurements.lineHeight*float64(i) + t.measurements.ascent))
+		if ly > screen.Bounds().Max.Y+int(math.Round(t.measurements.ascent)) {
```

Comment:
So I think screen is the wrong variable here, I was testing with the textarea example widget and found that this was never being hit. I think this needs to be played with more.

Reply:
Oh definitely, good catch! 🤦 
My use case being with a text area that takes up the whole height, I missed that…

Reply:
Actually, after looking again at the code, I’m not quite sure I understand what you mean.
I do test with the TextArea as well (see diff in the description) and it seems to be helping quite significantly.
We can only check the bounds of the screen, as even if in the end drawing the part of the text outside the ScrollContainer viewport is a no-op, the Text widget does not know of the mask set on the buffer in the ScrollContainer (assuming I correctly figured out how it works 😅).

But perhaps I’m misunderstanding your point here?

Reply:
Also when I put a breakpoint on the `return` on next line, and run the TextArea example with my patch from the description, it definitely gets hit 🤔

Reply:
So at no point while I am running this example https://github.com/ebitenui/ebitenui/blob/master/_examples/widget_demos/textarea/main.go
does the return line get hit. I think it should be based off the text widget's rect instead.  See the screenshot below:

![image](https://github.com/ebitenui/ebitenui/assets/13891542/bff044b7-8921-4722-83d9-2aa2e07761bf)

Reply:
And yes with the changes you made it would start drawing off the screen so the return would get hit in that case, which is fine, but I feel like we can be more aggressive with this change for when there are smaller amounts of lines within the text object.

Reply:
Ah yes, but the Text rect includes the space for the whole text, without bounding it to the visible part of the TextArea, and its height is way bigger than the screen height, e.g. with my changes to the example to add a lot more lines:

![image](https://github.com/ebitenui/ebitenui/assets/2386884/f22be59b-8771-4099-954f-8dec6ae08cee)

Reply:
Ok, so I've been playing with this a bit tonight and found a good solution. Under the two checks you made we add the following:

```
		if t.widget.parent != nil {
			if ly < t.widget.parent.Rect.Min.Y {
				continue
			}
			if ly-int(math.Round(t.measurements.lineHeight)) > t.widget.parent.Rect.Max.Y {
				return
			}
		}
```
And in textarea.go we add at line 261
```
	l.text.widget.parent = l.container.GetWidget()
```

Doing this allows the text object to know the bounding box of its parent. If the text goes outside the bounding box of its parent it doesn't get drawn.